### DISCUSSION TITLE
Component purl vs purlCoordinates

### DISCUSSION THREAD

**Discussion by malice00** [DependencyTrack/dependency-track#2330] [Q&A] [answered]
I was looking in the code to maybe fix some issues connected to the ALL-operator in policies, which led me to (among others) #1925.

I though this was maybe a good starting point, but now I do have some questions regarding the fields `purl` and `purlCoordinates` in the `org.dependencytrack.model.Component` class. _It looks to me (from looking through my database)_ that `purlCoordinates` is **always** a substring of (possibly even equal to) `purl`, and it **always** starts from position 0.

Now, my first question would be: is this correct, or are there cases where this is not valid? If so, then #1925 is indeed a valid issue and can be easily fixed. If not, then the question remains, if there is some valid point to #1925 and if and when 2 Viollations should actually be added to the list.

My second question arose when looking in the class `org.dependencytrack.policy.PackageURLPolicyEvaluator`, in which the method `evaluate` is aborted when the `purl` of the component is `null`: In case the fields can be different (Q1), is it correct to only check the `purl` when evaluating?

I'm hoping to get some answers, preferably after a little discussion, so I can 'get my hands dirty' and dive in.

Thanks!

**Comment by stevespringett** [comment, marked as answer]
Purl does two things. It can identify a component, and it can locate a component. Location, if not the default, is handled with qualifiers. This is the equivalent of a URL querystring. purlCordinates if the first part of the purl containing the ecosystem, optional namespace, name, and version. It does not include any qualifiers. The purpose for having both is to speed up database queries. If I was querying the db using a components purl identity, I can perform an exact match on purlCoordinates which is an indexed field. If I only had a single purl field, DT would have to perform a full table scan when performing the same query, which would negatively impact performance in a big way.

**Reply by malice00** [reply]
So it is (mostly?) a database-thing to have both fields? Does that mean that in code, it would be OK to work with just the purl?

**Reply by stevespringett** [reply]
> it would be OK to work with just the purl

Depends on what you're trying to do. If you need to query a component by its identity, I'd recommend always using purlCoordinates. For example, if you're querying for `pkg:maven/org.apache.logging.log4j/log4j-core@2.19.0` but the purl for it is `pkg:maven/org.apache.logging.log4j/log4j-core?type=jar`, then you won't get any results. So you'll need to use purlCoordinates this this case. which is what the policy evaluators, scanners, and dependency graphs do.

**Reply by malice00** [reply]
I'm referring to the lines mentioned in #1925:

https://github.com/DependencyTrack/dependency-track/blob/7685a29143dbb23f6d63af20ca4845790d81023b/src/main/java/org/dependencytrack/policy/PackageURLPolicyEvaluator.java#L58-L68

Seeing as here you are checking if the String values of both fields **contains** the configuration, I feel using only `purl` should be enough (seeing it is the 'more specific' of the two). Also, imho `purl` should be preferred in case the configuration actually uses the querystring to filter specific types or classifiers.

Do you agree it is OK to only check the `purl` here?

**Reply by stevespringett** [reply]
If you only use purl, you'll need to ask the database to perform a full table scan. This logic should be changed so that if either getPurl or getPurlCoordinates has a violation, then only one is added, not two. But both getPurl and getPurlCoordinates need to be in there.

**Reply by stevespringett** [reply]
Perhaps something like:
```java
if (component.getPurl() != null) { 
    if (component.getPurl().canonicalize().contains(condition.getValue())) { 
        violations.add(new PolicyConditionViolation(condition, component)); 
    } else {
        if (component.getPurlCoordinates() != null) {
            if (component.getPurlCoordinates().canonicalize().contains(condition.getValue())) { 
                violations.add(new PolicyConditionViolation(condition, component)); 
            } 
        }
    }
} 
```

It should not be possible to have purlCoordinates without also having a purl.

**Reply by malice00** [reply]
I'm not sure why a table-scan would be required in case of the policy-evaluation -- I assumed I already have the component that needs to be evaluated in hand (fully loaded) and in the code I see no searches being done... But I admit, I haven't had time yet to do more than discuss the issue with you here, so I might be missing something.

As or the code you posted and the comment you added: imho the else-block is totally useless -- if the purl doesn't match, the purlCoordinates will never match, seeing they are a substring of the purl... That's why I mentioned simply only using the purl to evaluate policies. If I'm wrong, could you maybe give me an example of a case where your code would actually be `false` on the first `if` but then return true on the last?

**Reply by stevespringett** [reply]
> As or the code you posted and the comment you added: imho the else-block is totally useless

You're right. Get rid of it to read:
```java
if (component.getPurl() != null) { 
    if (component.getPurl().canonicalize().contains(condition.getValue())) { 
        violations.add(new PolicyConditionViolation(condition, component)); 
    }
} 
```

**Reply by malice00** [reply]
Thanks, I'll implement this and some tests in the next couple of days!
Thank you for the back-and-forth, just want to make sure I understand the code and don't add anything stupid! :-)

**Reply by stevespringett** [reply]
And thanks for the contributions. Much appreciated.
